In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

<h3>Load bank data</h3>

In [68]:
from google.colab import drive
drive.mount('/content/gdrive',force_remount=True)

Mounted at /content/gdrive


In [69]:
bank_df = pd.read_csv('/content/gdrive/My Drive/AIML/Projects/DeepLearning/bank.csv')
bank_df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [70]:
bank_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             10000 non-null int64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


<h3>Analysis of data</h3>

To find out whether some of the columns like CustomerId and RowNumber are significant, finding out the value counts for the unique values as below. Seems like these two columns can be removed


*   CustomerId
*   RowNumber

There are 3 non-numeric columns, which can be label encoded and those are listed below:

*   Surname
*   Geography
*   Gender






---



In [71]:
bank_df['CustomerId'].value_counts().unique()

array([1])

In [72]:
bank_df['RowNumber'].value_counts().unique()

array([1])

In [0]:
bank_df.drop(columns=['RowNumber','CustomerId'], axis=1, inplace=True)

In [74]:
bank_df.head()

Surname  CreditScore Geography  Gender  Age  Tenure    Balance  \
0  Hargrave          619    France  Female   42       2       0.00   
1      Hill          608     Spain  Female   41       1   83807.86   
2      Onio          502    France  Female   42       8  159660.80   
3      Boni          699    France  Female   39       1       0.00   
4  Mitchell          850     Spain  Female   43       2  125510.82   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0              1          1               1        101348.88       1  
1              1          0               1        112542.58       0  
2              3          1               0        113931.57       1  
3              2          0               0         93826.63       0  
4              1          1               1         79084.10       0

<h4>Encode object values</h4>

In [75]:
bank_df['Surname'].value_counts().unique()

array([32, 29, 28, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13,
       12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1])

In [76]:
bank_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             10000 non-null int64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(2), int64(7), object(3)
memory usage: 937.6+ KB


In [0]:
from sklearn.preprocessing import LabelEncoder
le_surname = LabelEncoder()
le_geo = LabelEncoder()
le_gender = LabelEncoder()

In [78]:
bank_df['Surname'].value_counts().head()

Smith     32
Scott     29
Martin    29
Walker    28
Brown     26
Name: Surname, dtype: int64

In [79]:
bank_df['Geography'].value_counts()

France     5014
Germany    2509
Spain      2477
Name: Geography, dtype: int64

In [80]:
bank_df['Gender'].value_counts()

Male      5457
Female    4543
Name: Gender, dtype: int64

In [0]:
bank_df['Surname1'] = le_surname.fit_transform(bank_df['Surname'])
bank_df['Geography1'] = le_geo.fit_transform(bank_df['Geography'])
bank_df['Gender1'] = le_geo.fit_transform(bank_df['Gender'])

In [82]:
bank_df['Surname1'].value_counts().head()

2473    32
1689    29
2389    29
2751    28
336     26
Name: Surname1, dtype: int64

In [83]:
bank_df['Geography1'].value_counts()

0    5014
1    2509
2    2477
Name: Geography1, dtype: int64

In [84]:
bank_df['Gender1'].value_counts()

1    5457
0    4543
Name: Gender1, dtype: int64

In [85]:
bank_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 15 columns):
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             10000 non-null int64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
Surname1           10000 non-null int64
Geography1         10000 non-null int64
Gender1            10000 non-null int64
dtypes: float64(2), int64(10), object(3)
memory usage: 1.1+ MB


<h3>Dependant and Independant variables</h3>

Defining X with all the independant variables including the newly formed columns with encoded values for the 3 columns - Geography, Gender and Surname. Surname may be considered as non-significant, but it may have some impact on this data. Hence including the same

Defining Y with dependant variable - Exited

In [0]:
X = bank_df[['CreditScore','Age','Tenure','Balance','NumOfProducts','HasCrCard','IsActiveMember','EstimatedSalary','Surname1','Geography1','Gender1']]
Y = bank_df[['Exited']]

In [87]:
X.shape

(10000, 11)

In [88]:
Y.shape

(10000, 1)

<h3>Train-Test split</h3>

Train test split is done with 70:30 ratio and the shapes of each of the train test data for independant and dependant variables are printed below

In [0]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.3, random_state=7)

In [90]:
X_train.shape

(7000, 11)

In [91]:
X_test.shape

(3000, 11)

In [92]:
Y_train.shape

(7000, 1)

In [93]:
Y_test.shape

(3000, 1)

<h3>Normalization of data</h3>

All the variables are in different scales and hence, normalization is critical. Care is taken to normalize the data after train-test split to reduce the effect of test data in order to avoid bias errors

In [0]:
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()

In [95]:
X_train_ss = ss.fit_transform(X_train)
X_test_ss = ss.fit_transform(X_test)
#Y_train_ss = ss.fit_transform(Y_train)
#Y_test_ss = ss.fit_transform(Y_test)

/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/usr/local/lib/python3.6/dist-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/usr/local/lib/python3.6/dist-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [96]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7000 entries, 2317 to 9412
Data columns (total 11 columns):
CreditScore        7000 non-null int64
Age                7000 non-null int64
Tenure             7000 non-null int64
Balance            7000 non-null float64
NumOfProducts      7000 non-null int64
HasCrCard          7000 non-null int64
IsActiveMember     7000 non-null int64
EstimatedSalary    7000 non-null float64
Surname1           7000 non-null int64
Geography1         7000 non-null int64
Gender1            7000 non-null int64
dtypes: float64(2), int64(9)
memory usage: 656.2 KB


In [97]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3000 entries, 1977 to 3582
Data columns (total 11 columns):
CreditScore        3000 non-null int64
Age                3000 non-null int64
Tenure             3000 non-null int64
Balance            3000 non-null float64
NumOfProducts      3000 non-null int64
HasCrCard          3000 non-null int64
IsActiveMember     3000 non-null int64
EstimatedSalary    3000 non-null float64
Surname1           3000 non-null int64
Geography1         3000 non-null int64
Gender1            3000 non-null int64
dtypes: float64(2), int64(9)
memory usage: 281.2 KB


In [98]:
Y_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7000 entries, 2317 to 9412
Data columns (total 1 columns):
Exited    7000 non-null int64
dtypes: int64(1)
memory usage: 109.4 KB


In [99]:
Y_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3000 entries, 1977 to 3582
Data columns (total 1 columns):
Exited    3000 non-null int64
dtypes: int64(1)
memory usage: 46.9 KB


<h3>Logistic Regression</h3>

Logistic regression is performed here with scikit learn to find out the accuracy/confusion matrix/precision & recall score/f1-score of the model. Later this can be used to compare with the accuracy obtained using neural network


*   78.93% accuracy with logistic regression
*   F1-score of 88% for not exited use case and 8% for exited use case is obtained as shown below



In [100]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(X_train_ss,Y_train)
print(lr.score(X_test_ss,Y_test))
Y_pred = lr.predict(X_test_ss)
from sklearn.metrics import classification_report, confusion_matrix
print(classification_report(Y_test,Y_pred))
print(confusion_matrix(Y_test,Y_pred))


0.8116666666666666
              precision    recall  f1-score   support

           0       0.82      0.97      0.89      2395
           1       0.61      0.19      0.28       605

   micro avg       0.81      0.81      0.81      3000
   macro avg       0.72      0.58      0.59      3000
weighted avg       0.78      0.81      0.77      3000

[[2323   72]
 [ 493  112]]


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


<h3>Keras</h3>

<h4>Neural Network with no hidden layer</h4>

Neural network with no hidden layer has only input layer with 11 variables and 1 output/dependant variables. Default learning rate with batch value of 11 along with epochs of 50 is used

Sigmoid function gives values from 0 to 1 for the test set. With 0.5 as threshold, if the predicted value is > 0.5, output will be 1 and if the predicted value < 0.5, output will be 0

This Neural Network gives an accuracy of 81.3% and f1-score of 89% (true) and 30% (false) and 77% as weighted average

In [0]:
import keras
import tensorflow as tf
from sklearn.metrics import accuracy_score

<h5>Convert Y_pred to binary</h5>

Convert Y_pred to binary values based on 0.5 as threshold

In [0]:
def to_y_pred_bin(Y_pred_input):
  Y_pred_output = Y_pred_input
  for i in range(Y_pred_input.shape[0]):

          # Convert probabilities A[0,i] to actual predictions p[0,i]
          ### START CODE HERE ### (≈ 4 lines of code)
          if Y_pred_input[i ,0] >= 0.5:
              Y_pred_output[i ,0] = 1
          else:
              Y_pred_output[i,0] = 0
          ### END CODE HERE ###
  return Y_pred_output

In [103]:
#Initialize Sequential model
model = tf.keras.models.Sequential()

#Reshape data from 2D to 1D
model.add(tf.keras.layers.Reshape((11,),input_shape=(11,)))

#Normalize the data
model.add(tf.keras.layers.BatchNormalization())

#Add Dense Layer which provides 1 Output after applying sigmoid
model.add(tf.keras.layers.Dense(1, activation='sigmoid'))


#Compile the model
model.compile(optimizer='sgd', loss='binary_crossentropy', metrics=['accuracy'])

model.fit(X_train_ss, Y_train, 
          validation_data=(X_test_ss, Y_test), 
          epochs=50,
          batch_size=X_train_ss.shape[1])

Y_NoHidden_pred = model.predict(X_test_ss)
Z_NoHidden_pred = to_y_pred_bin(Y_NoHidden_pred)

print(accuracy_score(Y_test,Z_NoHidden_pred))
print(confusion_matrix(Y_test,Z_NoHidden_pred))
print(classification_report(Y_test,Z_NoHidden_pred))

Train on 7000 samples, validate on 3000 samples
Epoch 1/50
7000/7000 [==============================] - 2s 215us/sample - loss: 0.5673 - acc: 0.7204 - val_loss: 0.4868 - val_acc: 0.7750
Epoch 2/50
7000/7000 [==============================] - 1s 144us/sample - loss: 0.4627 - acc: 0.7900 - val_loss: 0.4516 - val_acc: 0.7937
Epoch 3/50
7000/7000 [==============================] - 1s 139us/sample - loss: 0.4480 - acc: 0.7981 - val_loss: 0.4397 - val_acc: 0.8043
Epoch 4/50
7000/7000 [==============================] - 1s 136us/sample - loss: 0.4399 - acc: 0.8056 - val_loss: 0.4370 - val_acc: 0.8067
Epoch 5/50
7000/7000 [==============================] - 1s 139us/sample - loss: 0.4402 - acc: 0.8083 - val_loss: 0.4347 - val_acc: 0.8107
Epoch 6/50
7000/7000 [==============================] - 1s 142us/sample - loss: 0.4354 - acc: 0.8101 - val_loss: 0.4350 - val_acc: 0.8050
Epoch 7/50
7000/7000 [==============================] - 1s 137us/sample - loss: 0.4373 - acc: 0.8069 - val_loss: 0.4343 - va

<h4>Neural Network with no hidden layer with learning rate of 0.03</h4>

Neural network with no hidden layer has only input layer with 11 variables and 1 output/dependant variables. Learning rate of 0.03/0.1 along with batch value of 11 with epochs as 100 is used

Sigmoid function gives values from 0 to 1 for the test set. With 0.5 as threshold, if the predicted value is > 0.5, output will be 1 and if the predicted value < 0.5, output will be 0

This Neural Network gives an accuracy of 81.3% and f1-score of 89% (true) and 30% (false) and 77% as weighted average

In [104]:
#Initialize Sequential model
model = tf.keras.models.Sequential()

#Reshape data from 2D to 1D
model.add(tf.keras.layers.Reshape((11,),input_shape=(11,)))

#Normalize the data
model.add(tf.keras.layers.BatchNormalization())

#Add Dense Layer which provides 1 Output after applying softmax
model.add(tf.keras.layers.Dense(1, activation='sigmoid'))


#Comile the model
sgd_optimizer = tf.keras.optimizers.SGD(lr=0.1)
model.compile(optimizer=sgd_optimizer, loss='binary_crossentropy', metrics=['accuracy'])

model.fit(X_train_ss, Y_train, 
          validation_data=(X_test_ss, Y_test), 
          epochs=50,
          batch_size=X_train_ss.shape[1])

Y_lr_pred = model.predict(X_test_ss)
Z_lr_pred = to_y_pred_bin(Y_lr_pred)

print(accuracy_score(Y_test,Z_lr_pred))
print(confusion_matrix(Y_test,Z_lr_pred))
print(classification_report(Y_test,Z_lr_pred))

Train on 7000 samples, validate on 3000 samples
Epoch 1/50
7000/7000 [==============================] - 2s 217us/sample - loss: 0.4713 - acc: 0.7896 - val_loss: 0.4363 - val_acc: 0.8080
Epoch 2/50
7000/7000 [==============================] - 1s 138us/sample - loss: 0.4437 - acc: 0.8069 - val_loss: 0.4351 - val_acc: 0.8063
Epoch 3/50
7000/7000 [==============================] - 1s 137us/sample - loss: 0.4422 - acc: 0.8057 - val_loss: 0.4355 - val_acc: 0.8113
Epoch 4/50
7000/7000 [==============================] - 1s 138us/sample - loss: 0.4403 - acc: 0.8073 - val_loss: 0.4322 - val_acc: 0.8117
Epoch 5/50
7000/7000 [==============================] - 1s 135us/sample - loss: 0.4390 - acc: 0.8106 - val_loss: 0.4384 - val_acc: 0.8020
Epoch 6/50
7000/7000 [==============================] - 1s 137us/sample - loss: 0.4408 - acc: 0.8056 - val_loss: 0.4337 - val_acc: 0.8103
Epoch 7/50
7000/7000 [==============================] - 1s 136us/sample - loss: 0.4385 - acc: 0.8073 - val_loss: 0.4342 - va

<h4>Optimization</h4>
<h4>Neural Network with  no hidden layer (batch value=52)</h4>

Neural network with no hidden layer has only input layer with 11 variables and 1 output/dependant variables. Default learning rate is used and batch value of 52 is used

Sigmoid function gives values from 0 to 1 for the test set. With 0.5 as threshold, if the predicted value is > 0.5, output will be 1 and if the predicted value < 0.5, output will be 0

This Neural Network gives an accuracy of 81.3% and f1-score of 89% (true) and 30% (false) and 77% as weighted average

In [105]:
#Initialize Sequential model
model = tf.keras.models.Sequential()

#Reshape data from 2D to 1D
model.add(tf.keras.layers.Reshape((11,),input_shape=(11,)))

#Normalize the data
model.add(tf.keras.layers.BatchNormalization())

#Add Dense Layer which provides 1 Output after applying softmax
model.add(tf.keras.layers.Dense(1, activation='sigmoid'))


#Comile the model
sgd_optimizer = tf.keras.optimizers.SGD(lr=0.03)
model.compile(optimizer="sgd", loss='binary_crossentropy', metrics=['accuracy'])

model.fit(X_train_ss,Y_train, validation_data=(X_test_ss,Y_test),
          epochs=100,batch_size=52)

Y_pred_batch = model.predict(X_test_ss)
Z_pred_batch = to_y_pred_bin(Y_pred_batch)

print(accuracy_score(Y_test,Y_pred_batch))
print(confusion_matrix(Y_test,Y_pred_batch))
print(classification_report(Y_test,Y_pred_batch))

Train on 7000 samples, validate on 3000 samples
Epoch 1/100
7000/7000 [==============================] - 1s 116us/sample - loss: 0.7347 - acc: 0.6084 - val_loss: 0.6333 - val_acc: 0.6850
Epoch 2/100
7000/7000 [==============================] - 0s 34us/sample - loss: 0.5873 - acc: 0.7123 - val_loss: 0.5543 - val_acc: 0.7393
Epoch 3/100
7000/7000 [==============================] - 0s 33us/sample - loss: 0.5357 - acc: 0.7487 - val_loss: 0.5181 - val_acc: 0.7607
Epoch 4/100
7000/7000 [==============================] - 0s 33us/sample - loss: 0.5082 - acc: 0.7604 - val_loss: 0.4979 - val_acc: 0.7700
Epoch 5/100
7000/7000 [==============================] - 0s 34us/sample - loss: 0.4913 - acc: 0.7720 - val_loss: 0.4843 - val_acc: 0.7777
Epoch 6/100
7000/7000 [==============================] - 0s 34us/sample - loss: 0.4810 - acc: 0.7777 - val_loss: 0.4747 - val_acc: 0.7797
Epoch 7/100
7000/7000 [==============================] - 0s 34us/sample - loss: 0.4728 - acc: 0.7827 - val_loss: 0.4673 - v

<h4>Optimization</h4>
<h4>Neural Network with  1 hidden layer (relu function)</h4>

Neural network with 1 hidden layer has only input layer with 11 variables and 1 output/dependant variables. Learning rate of 0.03 is used and batch value of 52 is used

Sigmoid function gives values from 0 to 1 for the test set. With 0.5 as threshold, if the predicted value is > 0.5, output will be 1 and if the predicted value < 0.5, output will be 0

This Neural Network gives an accuracy of 85.73% and f1-score of 92% (true) and 54% (false) and 84% as weighted average

In [106]:
#Initialize Sequential model
model = tf.keras.models.Sequential()

#Reshape data from 2D to 1D
model.add(tf.keras.layers.Reshape((11,),input_shape=(11,)))

#Normalize the data
model.add(tf.keras.layers.BatchNormalization())

#Add 1st hidden layer
model.add(tf.keras.layers.Dense(10, activation='relu'))

#Add OUTPUT layer
model.add(tf.keras.layers.Dense(1, activation='sigmoid'))

#Create optimizer with non-default learning rate
sgd_optimizer = tf.keras.optimizers.SGD(lr=0.03)

#Compile the model
model.compile(optimizer=sgd_optimizer, loss='binary_crossentropy', metrics=['accuracy'])

model.fit(X_train_ss, Y_train, validation_data=(X_test_ss,Y_test), epochs=50, batch_size=52)

Y_1_pred = model.predict(X_test_ss)
Z_1_pred = to_y_pred_bin(Y_1_pred)

print(accuracy_score(Y_test,Z_1_pred))
print(confusion_matrix(Y_test,Z_1_pred))
print(classification_report(Y_test,Z_1_pred))

Train on 7000 samples, validate on 3000 samples
Epoch 1/50
7000/7000 [==============================] - 1s 127us/sample - loss: 0.5471 - acc: 0.7620 - val_loss: 0.5046 - val_acc: 0.7873
Epoch 2/50
7000/7000 [==============================] - 0s 36us/sample - loss: 0.4905 - acc: 0.7907 - val_loss: 0.4751 - val_acc: 0.7927
Epoch 3/50
7000/7000 [==============================] - 0s 35us/sample - loss: 0.4716 - acc: 0.7923 - val_loss: 0.4605 - val_acc: 0.7933
Epoch 4/50
7000/7000 [==============================] - 0s 36us/sample - loss: 0.4598 - acc: 0.7913 - val_loss: 0.4513 - val_acc: 0.7933
Epoch 5/50
7000/7000 [==============================] - 0s 38us/sample - loss: 0.4513 - acc: 0.7921 - val_loss: 0.4450 - val_acc: 0.7980
Epoch 6/50
7000/7000 [==============================] - 0s 36us/sample - loss: 0.4473 - acc: 0.7947 - val_loss: 0.4409 - val_acc: 0.7980
Epoch 7/50
7000/7000 [==============================] - 0s 36us/sample - loss: 0.4440 - acc: 0.7956 - val_loss: 0.4372 - val_acc:

<h4>Optimization</h4>
<h4>Neural Network with 1 hidden layer (tanh function)</h4>

Neural network with 1 hidden layer (with tanh activation function) has input layer with 11 variables and 1 output/dependant variables. Learning rate of 0.03 is used and batch value of 52 is used

Sigmoid function gives values from 0 to 1 for the test set. With 0.5 as threshold, if the predicted value is > 0.5, output will be 1 and if the predicted value < 0.5, output will be 0

This Neural Network gives an accuracy of 85.83% and f1-score of 92% (true) and 54% (false) and 84% as weighted average

In [107]:
#Initialize Sequential model
model = tf.keras.models.Sequential()

#Reshape data from 2D to 1D
model.add(tf.keras.layers.Reshape((11,),input_shape=(11,)))

#Normalize the data
model.add(tf.keras.layers.BatchNormalization())

#Add 1st hidden layer (tanh)
model.add(tf.keras.layers.Dense(10, activation='tanh'))

#Add OUTPUT layer
model.add(tf.keras.layers.Dense(1, activation='sigmoid'))

#Create optimizer with non-default learning rate
sgd_optimizer = tf.keras.optimizers.SGD(lr=0.03)

#Compile the model
model.compile(optimizer=sgd_optimizer, loss='binary_crossentropy', metrics=['accuracy'])

model.fit(X_train_ss,Y_train,validation_data=(X_test_ss,Y_test),epochs=100,batch_size=52)

Y_tan_pred = model.predict(X_test_ss)
Z_tan_pred = to_y_pred_bin(Y_tan_pred)

print(accuracy_score(Y_test,Z_tan_pred))
print(classification_report(Y_test,Z_tan_pred))
print(confusion_matrix(Y_test,Z_tan_pred))

Train on 7000 samples, validate on 3000 samples
Epoch 1/100
7000/7000 [==============================] - 1s 127us/sample - loss: 0.5797 - acc: 0.7100 - val_loss: 0.4807 - val_acc: 0.7960
Epoch 2/100
7000/7000 [==============================] - 0s 39us/sample - loss: 0.4640 - acc: 0.7964 - val_loss: 0.4475 - val_acc: 0.8030
Epoch 3/100
7000/7000 [==============================] - 0s 36us/sample - loss: 0.4449 - acc: 0.7997 - val_loss: 0.4365 - val_acc: 0.8080
Epoch 4/100
7000/7000 [==============================] - 0s 38us/sample - loss: 0.4371 - acc: 0.8007 - val_loss: 0.4309 - val_acc: 0.8077
Epoch 5/100
7000/7000 [==============================] - 0s 36us/sample - loss: 0.4330 - acc: 0.8063 - val_loss: 0.4267 - val_acc: 0.8123
Epoch 6/100
7000/7000 [==============================] - 0s 36us/sample - loss: 0.4300 - acc: 0.8073 - val_loss: 0.4231 - val_acc: 0.8140
Epoch 7/100
7000/7000 [==============================] - 0s 36us/sample - loss: 0.4231 - acc: 0.8090 - val_loss: 0.4195 - v

<h4>Optimization</h4>
<h4>Neural Network with 1 hidden layer (leaky relu function)</h4>

Neural network with 1 hidden layer (with leaky relu activation function) has input layer with 11 variables and 1 output/dependant variables. Learning rate of 0.03 is used and batch value of 52 is used

Sigmoid function gives values from 0 to 1 for the test set. With 0.5 as threshold, if the predicted value is > 0.5, output will be 1 and if the predicted value < 0.5, output will be 0

This Neural Network gives an accuracy of 83.13% and f1-score of 90% (true) and 40% (false) and 80% as weighted average

In [108]:
#Initialize Sequential model
model = tf.keras.models.Sequential()

#Reshape data from 2D to 1D
model.add(tf.keras.layers.Reshape((11,),input_shape=(11,)))

#Normalize the data
model.add(tf.keras.layers.BatchNormalization())

#Add 1st hidden layer (tanh)
model.add(tf.keras.layers.LeakyReLU(alpha=0.03))

#Add OUTPUT layer
model.add(tf.keras.layers.Dense(1, activation='sigmoid'))

#Create optimizer with non-default learning rate
sgd_optimizer = tf.keras.optimizers.SGD(lr=0.03)

#Compile the model
model.compile(optimizer=sgd_optimizer, loss='binary_crossentropy', metrics=['accuracy'])

model.fit(X_train_ss,Y_train,validation_data=(X_test_ss,Y_test),epochs=150,batch_size=52)

Y_prelu_pred = model.predict(X_test_ss)
Z_prelu_pred = to_y_pred_bin(Y_prelu_pred)


print(accuracy_score(Y_test,Z_prelu_pred))
print(classification_report(Y_test,Z_prelu_pred))
print(confusion_matrix(Y_test,Z_prelu_pred))

Train on 7000 samples, validate on 3000 samples
Epoch 1/150
7000/7000 [==============================] - 1s 128us/sample - loss: 0.5447 - acc: 0.7803 - val_loss: 0.5144 - val_acc: 0.7960
Epoch 2/150
7000/7000 [==============================] - 0s 36us/sample - loss: 0.5103 - acc: 0.7950 - val_loss: 0.4951 - val_acc: 0.7983
Epoch 3/150
7000/7000 [==============================] - 0s 35us/sample - loss: 0.4961 - acc: 0.7956 - val_loss: 0.4831 - val_acc: 0.7987
Epoch 4/150
7000/7000 [==============================] - 0s 36us/sample - loss: 0.4841 - acc: 0.7974 - val_loss: 0.4728 - val_acc: 0.8020
Epoch 5/150
7000/7000 [==============================] - 0s 36us/sample - loss: 0.4736 - acc: 0.8013 - val_loss: 0.4634 - val_acc: 0.8083
Epoch 6/150
7000/7000 [==============================] - 0s 36us/sample - loss: 0.4655 - acc: 0.8066 - val_loss: 0.4559 - val_acc: 0.8103
Epoch 7/150
7000/7000 [==============================] - 0s 35us/sample - loss: 0.4588 - acc: 0.8054 - val_loss: 0.4511 - v

<h4>Optimization</h4>
<h4>Neural Network with 2 hidden layer</h4>

Neural network with 2 hidden layers has input layer with 11 variables and 1 output/dependant variables. Learning rate of 0.03 is used and batch value of 52 is used

Sigmoid function gives values from 0 to 1 for the test set. With 0.5 as threshold, if the predicted value is > 0.5, output will be 1 and if the predicted value < 0.5, output will be 0

This Neural Network gives an accuracy of 85.73% and f1-score of 92% (true) and 54% (false) and 84% as weighted average

In [109]:
#Initialize Sequential model
model = tf.keras.models.Sequential()

#Reshape data from 2D to 1D 
model.add(tf.keras.layers.Reshape((11,),input_shape=(11,)))

#Normalize the data
model.add(tf.keras.layers.BatchNormalization())

#Add 1st hidden layer
model.add(tf.keras.layers.Dense(10, activation='relu'))##use sigmoid instead of relu for this residency

#Add 2nd hidden layer
model.add(tf.keras.layers.Dense(10, activation='tanh'))

#Add OUTPUT layer
model.add(tf.keras.layers.Dense(1, activation='sigmoid'))

#Create optimizer with non-default learning rate
sgd_optimizer = tf.keras.optimizers.SGD(lr=0.03)

#Compile the model
model.compile(optimizer=sgd_optimizer, loss='binary_crossentropy', metrics=['accuracy'])

print(model.summary())

model.fit(X_train_ss,Y_train,          
          validation_data=(X_test_ss,Y_test),
          epochs=100,
          batch_size=52)

Y_2_pred = model.predict(X_test_ss)
Z_2_pred = to_y_pred_bin(Y_2_pred)

print(accuracy_score(Y_test,Z_2_pred))
print(confusion_matrix(Y_test,Z_2_pred))
print(classification_report(Y_test,Z_2_pred))

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
reshape_23 (Reshape)         (None, 11)                0         
_________________________________________________________________
batch_normalization_v1_23 (B (None, 11)                44        
_________________________________________________________________
dense_33 (Dense)             (None, 10)                120       
_________________________________________________________________
dense_34 (Dense)             (None, 10)                110       
_________________________________________________________________
dense_35 (Dense)             (None, 1)                 11        
Total params: 285
Trainable params: 263
Non-trainable params: 22
_________________________________________________________________
None
Train on 7000 samples, validate on 3000 samples
Epoch 1/100
7000/7000 [==============================] - 1s 141us/sample - loss: 0.5184 - 

<h3>Summary</h3>

In this exercise, bank data were used to predict whether a user will exit or not. Bank data had 2 serial numbers which can be dropped as it does not seem to contribute to the prediction whether the customer will exit or not. The dataset also had 3 object datatypes, hence it was label encoded to convert to numerical values. Independant variables are formed with 11 datapoints. Dependant variable is formed with 1 datapoint. Dataset is split into train and test dataset for both independant and dependant variables. StandardScalar is applied to normalize the data

*   Logistic Regression is done using scikit to find out the initial accuracy (81%) along with f1-score (positive) (89%) and f1-score (negative) (28%) and overall weighted avg is 77%
*   NeuralNetwork is developed with no hidden layer with default learning rate and learning rate = 0.03 with batch size = 11 and 52 gives the same accuracy as the one given by logistic regression
*   <b><u>Optimization</u></b>
*   NeuralNetwork is developed with one hidden layer of 10 nodes with relu activation function with learning rate = 0.03 and with batch size = 52 gives an accuracy of 85.14% along with f1-score(positive) = 91% and f1-score (negative) = 53% and overall weighted avg is 84%
*   NeuralNetwork is developed with one hidden layer of 10 nodes with tanh activation function with learning rate = 0.03 and with batch size = 52 gives an accuracy of 86% along with f1-score(positive) = 92% and f1-score (negative) = 54% and overall weighted avg is 84%
*   NeuralNetwork is developed with one hidden layer of 10 nodes with leaky relu activation function with learning rate = 0.03 and with batch size = 52 gives an accuracy of 83.13% along with f1-score(positive) = 90% and f1-score (negative) = 40% and overall weighted avg is 80% 
*   NeuralNetwork is developed with two hidden layers of 10 nodes each one with relu and one with tanh activation function with learning rate = 0.03 and with batch size = 52 gives an accuracy of 85.6% along with f1-score(positive) = 91% and f1-score(negative) = 55% and overall weighted avg is 84%



